README.rst
*************************
django-system-maintenance
*************************

``django-system-maintenance`` is a Django app to document and track the administration and maintenance of computer systems.

Source code is available on GitHub at `mfcovington/django-system-maintenance <https://github.com/mfcovington/django-system-maintenance>`_.

.. contents:: :local:


Installation
============

**PyPI**

.. code-block:: sh

    pip install django-system-maintenance


**GitHub (development branch)**

.. code-block:: sh

    pip install git+http://github.com/mfcovington/django-system-maintenance.git@develop


Configuration
=============

Add ``system_maintenance`` and its dependencies to ``INSTALLED_APPS`` in ``settings.py``:

.. code-block:: python

    INSTALLED_APPS = (
        ...
        'django.contrib.humanize',
        'markupfield_helpers',
        'project_home_tags',
        'system_maintenance',
    )

Add the ``system_maintenance`` URLs to the site's ``urls.py``:

.. code-block:: python

    from django.urls import include, path


    urlpatterns = [
        ...
        path('system_maintenance/', include('system_maintenance.urls', namespace='system_maintenance')),
    ]


By default, lists of maintenance records, etc. are paginated with 30 records per page. This value can be customized in ``settings.py``:

.. code-block:: python

    SYSTEM_MAINTENANCE_PAGINATE_BY = 50

This app is compatible with ``django-project-home-templatetags``. Check out its `Configuration Documentation <https://github.com/mfcovington/django-project-home-templatetags#configuration>`_ if you want this app's top-level breadcrumb to link to your project's homepage. To activate ``project_home_tags`` functionality, you must define ``PROJECT_HOME_NAMESPACE`` and, optionally, ``PROJECT_HOME_LABEL`` in ``settings.py``:

.. code-block:: python

    PROJECT_HOME_NAMESPACE = 'project_name:index_view'    # Namespace of homepage
    PROJECT_HOME_LABEL = 'Homepage'    # Optional; Default is 'Home'


Migrations
==========

Create migrations for ``system_maintenance``, if necessary:

.. code-block:: sh

    python manage.py makemigrations system_maintenance


Perform migrations for ``system_maintenance``:

.. code-block:: sh

    python manage.py migrate


Usage
=====

- Start the development server:

.. code-block:: sh

    python manage.py runserver


- Login and add yourself as a system administrator: ``http://localhost:8000/admin/system_maintenance/sysadmin/add/``
- Visit: ``http://127.0.0.1:8000/system_maintenance/``


*Version 0.4.6*


CHANGELOG.rst
Revision History
================

0.4.6 2018-12-14

- Add missing migration


0.4.5 2018-12-13

- Remove ``MarkupField`` args already defined by markupfield helper


0.4.4 2018-12-06

- DRY up with markupfield_helpers


0.4.3 2018-12-06

- Include migrations when packaging for distribution


0.4.2 2018-11-26

- Add ``django-project-home-templatetags`` compatibility for breadcrumbs
- Include tests in distribution


0.4.1 2018-11-05

- Make compatible with Django 2.1
- Python>=3.4; Django>=2.0,<3.0


0.4.0 2018-11-05

- Make compatible with Django 2.0
- Drop Python 3.2 and 3.3 compatibility
- Python>=3.4; Django>=2.0,<=2.0.9


0.3.2 2018-11-05

- Set upper bound of <2.0 for Django version
- Python>=3.2,<3.8; Django>=1.7,<2.0
- Final version that will be compatible with Django 1.11.16


0.3.1 2018-11-05

- Prepare for upgrade to Django 2+
- Update dependency version requirements


0.3.0 2018-11-04

- Add unit tests and functional tests
- Add links to view raw markdown
- Update configuration instructions in README
- Set next page to System Maintenance home page if accessing authentication page directly
- Fix maintenance record status so it defaults to 'In Progress'
- Resolve Django 1.10 deprecation warnings
- DRY and simplify


0.2.0 2015-12-12

- Redirect to sysadmin authentication page instead of 404 if current user is not a sysadmin
- Make pagination customizable via ``settings.SYSTEM_MAINTENANCE_PAGINATE_BY``
- Open System Maintenance admin page in a new tab
- Add missing imports for plain text markup fields
- Add installation instructions to README
- Add default app configuration


0.1.0 2015-11-02

- A Django app to document and track the administration and maintenance of computer systems


system_maintenance/migrations/0001_initial.py
# Generated by Django 2.0.9 on 2018-11-05 10:44

from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion
import django.utils.timezone
import markupfield.fields


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='DocumentationRecord',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('title', models.CharField(help_text='Enter a brief, descriptive title for this documentation.', max_length=255, unique=True)),
                ('documentation', markupfield.fields.MarkupField(blank=True, help_text='Document how to perform a task.', null=True, rendered_field=True)),
                ('documentation_markup_type', models.CharField(choices=[('', '--'), ('Markdown', 'Markdown'), ('Markdown Basic', 'Markdown Basic'), ('Plain Text', 'Plain Text'), ('reStructuredText', 'reStructuredText')], default='Markdown', max_length=30)),
                ('created_at', models.DateTimeField(auto_now_add=True)),
                ('_documentation_rendered', models.TextField(editable=False, null=True)),
                ('updated_at', models.DateTimeField(auto_now=True)),
            ],
            options={
                'ordering': ['title'],
            },
        ),
        migrations.CreateModel(
            name='Hardware',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(help_text='Enter the type of hardware.', max_length=255, unique=True)),
            ],
            options={
                'verbose_name_plural': 'hardware',
                'ordering': ['name'],
            },
        ),
        migrations.CreateModel(
            name='MaintenanceRecord',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('description', markupfield.fields.MarkupField(blank=True, help_text='Enter a description of the system maintenance performed.', null=True, rendered_field=True)),
                ('description_markup_type', models.CharField(choices=[('', '--'), ('Markdown', 'Markdown'), ('Markdown Basic', 'Markdown Basic'), ('Plain Text', 'Plain Text'), ('reStructuredText', 'reStructuredText')], default='Markdown', max_length=30)),
                ('procedure', markupfield.fields.MarkupField(blank=True, help_text='Enter details of how the system maintenance was performed.', null=True, rendered_field=True)),
                ('_description_rendered', models.TextField(editable=False, null=True)),
                ('procedure_markup_type', models.CharField(choices=[('', '--'), ('Markdown', 'Markdown'), ('Markdown Basic', 'Markdown Basic'), ('Plain Text', 'Plain Text'), ('reStructuredText', 'reStructuredText')], default='Markdown', max_length=30)),
                ('problems', markupfield.fields.MarkupField(blank=True, help_text='Describe problems that arose during system maintenance.', null=True, rendered_field=True)),
                ('_procedure_rendered', models.TextField(editable=False, null=True)),
                ('problems_markup_type', models.CharField(choices=[('', '--'), ('Markdown', 'Markdown'), ('Markdown Basic', 'Markdown Basic'), ('Plain Text', 'Plain Text'), ('reStructuredText', 'reStructuredText')], default='Markdown', max_length=30)),
                ('_problems_rendered', models.TextField(editable=False, null=True)),
                ('status', models.CharField(choices=[('Complete', 'Complete'), ('In Progress', 'In Progress'), ('Failed', 'Failed')], default='In Progress', help_text='What is the current status of the system maintenance?', max_length=15)),
                ('datetime', models.DateTimeField(default=django.utils.timezone.now, help_text='Specify the date/time that the system maintenance was performed.', verbose_name='maintenance date/time')),
                ('documentation_records', models.ManyToManyField(blank=True, help_text='Select documentation relevant to this system maintenance.<br>', related_name='maintenance_records', to='system_maintenance.DocumentationRecord')),
                ('hardware', models.ManyToManyField(blank=True, help_text='Select the hardware involved in the system maintenance.', to='system_maintenance.Hardware')),
            ],
            options={
                'verbose_name': 'maintenance record',
                'verbose_name_plural': 'maintenance records',
                'ordering': ['-datetime'],
            },
        ),
        migrations.CreateModel(
            name='MaintenanceRecordRelationship',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('referenced_record', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='referenced_record', to='system_maintenance.MaintenanceRecord')),
                ('referencing_record', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='referencing_record', to='system_maintenance.MaintenanceRecord')),
            ],
            options={
                'ordering': ['referencing_record'],
            },
        ),
        migrations.CreateModel(
            name='MaintenanceType',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('maintenance_type', models.CharField(help_text="Enter a type of maintenance (e.g., 'Software Installation').", max_length=255, unique=True)),
                ('description', models.TextField(blank=True, help_text='Enter a description of the maintenance type.')),
            ],
            options={
                'ordering': ['maintenance_type'],
            },
        ),
        migrations.CreateModel(
            name='Software',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(help_text="Enter the software's name.", max_length=255, unique=True)),
            ],
            options={
                'verbose_name': 'software package',
                'verbose_name_plural': 'software packages',
                'ordering': ['name'],
            },
        ),
        migrations.CreateModel(
            name='SysAdmin',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('user', models.OneToOneField(help_text='Select a user.', on_delete=django.db.models.deletion.PROTECT, to=settings.AUTH_USER_MODEL)),
            ],
            options={
                'verbose_name': 'system administrator',
                'verbose_name_plural': 'system administrators',
            },
        ),
        migrations.CreateModel(
            name='System',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(help_text='Enter a brief, unique identifier for the system.', max_length=255, unique=True, verbose_name='system name')),
                ('description', models.TextField(blank=True, help_text='Enter a description of the system.')),
            ],
            options={
                'ordering': ['name'],
            },
        ),
        migrations.AddField(
            model_name='maintenancerecord',
            name='maintenance_type',
            field=models.ForeignKey(help_text='Select/Create a maintenance type.', on_delete=django.db.models.deletion.PROTECT, to='system_maintenance.MaintenanceType'),
        ),
        migrations.AddField(
            model_name='maintenancerecord',
            name='referenced_records',
            field=models.ManyToManyField(related_name='referencing_records', through='system_maintenance.MaintenanceRecordRelationship', to='system_maintenance.MaintenanceRecord'),
        ),
        migrations.AddField(
            model_name='maintenancerecord',
            name='software',
            field=models.ManyToManyField(blank=True, help_text='Select the software involved in the system maintenance.', to='system_maintenance.Software'),
        ),
        migrations.AddField(
            model_name='maintenancerecord',
            name='sys_admin',
            field=models.ForeignKey(help_text='Select a system administrator.', on_delete=django.db.models.deletion.PROTECT, to='system_maintenance.SysAdmin'),
        ),
        migrations.AddField(
            model_name='maintenancerecord',
            name='system',
            field=models.ForeignKey(help_text='Select/Create a system.', on_delete=django.db.models.deletion.PROTECT, to='system_maintenance.System'),
        ),
        migrations.AddField(
            model_name='documentationrecord',
            name='maintenance_type',
            field=models.ForeignKey(help_text='Select/Create a maintenance type.', on_delete=django.db.models.deletion.PROTECT, to='system_maintenance.MaintenanceType'),
        ),
        migrations.AlterUniqueTogether(
            name='maintenancerecordrelationship',
            unique_together={('referencing_record', 'referenced_record')},
        ),
    ]
